### PR TITLE
[NG] textarea component

### PR DESCRIPTION
**Pull request by gnomeontherun**
This introduces the new clrTextarea component, which handles the form validations for the textarea. It also extracts the control class logic into a service since it was being duplicated.

#1878

### CODE REVIEW

File: src/clr-angular/forms/textarea/textarea.spec.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+      it('should handle blur events', () => {
+        // textareas must be both invalid and blurred to register the validity
+        textarea.nativeElement.value = 'abc';
+        textarea.nativeElement.dispatchEvent(new Event('textarea'));
+        textarea.nativeElement.dispatchEvent(new Event('blur'));
+        fixture.detectChanges();
+        expect(ifErrorService.triggerStatusChange).toHaveBeenCalled();
+      });
+    });
```
Comment (youdz) [inline src/clr-angular/forms/textarea/textarea.spec.ts]: You're missing a test checking that it properly sets `controlClassService.className`.

File: src/clr-angular/forms/textarea/textarea-container.spec.ts [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+        });
+        fixture = TestBed.createComponent(NoLabelTest);
+
+        textareaContainerDE = fixture.debugElement.query(By.directive(ClrTextareaContainer));
+        textareaContainerEl = textareaContainerDE.nativeElement;
+        layoutService = textareaContainerDE.injector.get(LayoutService);
+      });
+
+      it('adds an empty label when instantiated without vertical layout', () => {
+        layoutService.layout = Layouts.HORIZONTAL;
```
Comment (youdz) [inline src/clr-angular/forms/textarea/textarea-container.spec.ts]: Maybe we should also test that the label **isn't** here when using a vertical layout?

File: src/clr-angular/forms/textarea/textarea-container.spec.ts [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        textareaContainerDE = fixture.debugElement.query(By.directive(ClrTextareaContainer));
+        textareaContainer = textareaContainerDE.componentInstance;
+        textareaContainerEl = textareaContainerDE.nativeElement;
+        ifErrorService = textareaContainerDE.injector.get(IfErrorService);
+        layoutService = textareaContainerDE.injector.get(LayoutService);
+        fixture.detectChanges();
+      });
+
+      it('injects the NgControlService and subscribes in AfterContentInit', () => {
```
Comment (youdz) [inline src/clr-angular/forms/textarea/textarea-container.spec.ts]: The description is off, it's the `IfErrorService ` there is no `AfterContentInit` involved.

File: src/clr-angular/forms/textarea/textarea-container.spec.ts [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+
+      it('computes the grid class for the control container when not vertical', () => {
+        expect(textareaContainer.controlClass()).toEqual('');
+        layoutService.layout = Layouts.HORIZONTAL;
+        expect(textareaContainer.controlClass()).toContain('clr-col-xs-12');
+      });
+
+      it('tracks the validity of the form control', () => {
+        let control;
+        const sub = ifErrorService.statusChanges.subscribe(ctrl => (control = ctrl));
```
Comment (youdz) [inline src/clr-angular/forms/textarea/textarea-container.spec.ts]: I don't understand this subscription, you're not using `control` at all in this test.

File: src/clr-angular/forms/textarea/textarea.ts [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+      );
+    }
+    if (controlClassService) {
+      controlClassService.className = el.nativeElement.className;
+    }
+  }
+
+  ngAfterContentInit() {
+    if (this.ngControlService) {
+      this.ngControlService.setControl(this.control);
```
[the remark this answers is not in the document]
Comment (gnomeontherun) [reply src/clr-angular/forms/textarea/textarea.ts]: Yeah, I looked into this again and I think there _was_ a reason before the final design and now its not needed. Moving up to constructor.

File: src/clr-angular/forms/textarea/textarea.spec.ts (hunk 1 above)
Comment (gnomeontherun) [reply src/clr-angular/forms/textarea/textarea.spec.ts]: Added

File: src/clr-angular/forms/textarea/textarea-container.spec.ts (hunk 4 above)
Comment (gnomeontherun) [reply src/clr-angular/forms/textarea/textarea-container.spec.ts]: Removed

File: src/clr-angular/forms/textarea/textarea-container.spec.ts (hunk 3 above)
Comment (gnomeontherun) [reply src/clr-angular/forms/textarea/textarea-container.spec.ts]: Fixed

File: src/clr-angular/forms/textarea/textarea-container.spec.ts (hunk 2 above)
Comment (gnomeontherun) [reply src/clr-angular/forms/textarea/textarea-container.spec.ts]: Done

**Comment by gnomeontherun** [comment]
@youdz I decided to consolidate the tests for input and textarea because they are practically identical. So the changeset is larger but reduces overall code (and will be better for future components).

**Pull request merged**